File: banchus-app/src/main/java/pe/nanamochi/banchus/service/RegistrationService.java
package pe.nanamochi.banchus.service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import pe.nanamochi.banchus.database.entity.User;
import pe.nanamochi.banchus.domain.enums.CountryCode;
import pe.nanamochi.banchus.domain.enums.ServerPrivileges;
import pe.nanamochi.banchus.util.Security;
import pe.nanamochi.banchus.util.Validation;

@Service
@RequiredArgsConstructor
public class RegistrationService {
  private final UserService userService;
  private final StatService statService;

  @Transactional
  public Map<String, List<String>> registerUser(
      String username, String email, String passwordPlainText, int check) {
    Map<String, List<String>> errors = new HashMap<>();

    if (check == 0) {
      if (!Validation.isValidUsername(username)) {
        addError(errors, "username", "Invalid username.");
      }
      if (!Validation.isValidEmail(email)) {
        addError(errors, "user_email", "Invalid email syntax.");
      }
      if (!Validation.isValidPassword(passwordPlainText)) {
        addError(
            errors,
            "password",
            "Password must be between 8 and 32 characters and contain more than 3 unique"
                + " characters.");
      }

      userService
          .findByUsername(username)
          .ifPresent(
              u -> addError(errors, "username", "Username already taken by another player."));

      userService
          .findByEmail(email)
          .ifPresent(u -> addError(errors, "user_email", "Email already taken by another player."));

      if (!errors.isEmpty()) return errors;

      User createdUser =
          userService.create(
              User.builder()
                  .username(username)
                  .email(email)
                  .passwordMd5(Security.getMd5(passwordPlainText))
                  .country(CountryCode.KP) // TODO: Default to North Korea for now
                  .privileges(ServerPrivileges.UNRESTRICTED.getValue())
                  .build());
      statService.createAllGamemodes(createdUser);
    }

    return Map.of();
  }

  private void addError(Map<String, List<String>> errors, String field, String message) {
    errors.computeIfAbsent(field, k -> new ArrayList<>()).add(message);
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/service/RegistrationService.kt
package pe.nanamochi.banchus.service

import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.binding
import com.github.michaelbull.result.mapError
import org.slf4j.LoggerFactory
import org.springframework.http.HttpHeaders
import org.springframework.security.crypto.bcrypt.BCrypt
import org.springframework.stereotype.Service
import pe.nanamochi.banchus.database.entity.User
import pe.nanamochi.banchus.domain.enums.CountryCode
import pe.nanamochi.banchus.domain.enums.ServerPrivileges
import pe.nanamochi.banchus.domain.error.CheckOk
import pe.nanamochi.banchus.domain.error.DomainMessage
import pe.nanamochi.banchus.domain.error.EmailTaken
import pe.nanamochi.banchus.domain.error.InvalidEmail
import pe.nanamochi.banchus.domain.error.InvalidPassword
import pe.nanamochi.banchus.domain.error.InvalidUsername
import pe.nanamochi.banchus.domain.error.RegistrationResult
import pe.nanamochi.banchus.domain.error.UserCreated
import pe.nanamochi.banchus.domain.error.UsernameTaken
import pe.nanamochi.banchus.util.isValidEmail
import pe.nanamochi.banchus.util.isValidPassword
import pe.nanamochi.banchus.util.isValidUsername
import pe.nanamochi.banchus.util.toMd5

@Service
class RegistrationService(
    private val userService: UserService,
    private val statService: StatService,
    private val geolocationService: GeolocationService,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun registerUser(
        username: String,
        passwordPlainText: String,
        email: String,
        check: Int,
        headers: HttpHeaders,
    ): Result<RegistrationResult, List<DomainMessage>> {
        if (check != 0) return Ok(CheckOk)

        val errors = buildList {
            if (!username.isValidUsername()) add(InvalidUsername)
            if (!email.isValidEmail()) add(InvalidEmail)
            if (!passwordPlainText.isValidPassword()) add(InvalidPassword)

            if (userService.fetchOneByUsername(username).isOk) add(UsernameTaken)
            if (userService.fetchOneByEmail(email).isOk) add(EmailTaken)
        }

        if (errors.isNotEmpty()) return Err(errors)

        val (_, geolocation) = geolocationService.resolve(headers)

        return binding {
                val createdUser =
                    userService
                        .create(
                            User(
                                username = username,
                                safeUsername = username.replace(" ", "_"),
                                email = email,
                                passwordBcrypt =
                                    BCrypt.hashpw(passwordPlainText.toMd5(), BCrypt.gensalt()),
                                country = CountryCode.fromCode(geolocation.countryCode),
                                privileges = ServerPrivileges.UNRESTRICTED.value,
                            )
                        )
                        .bind()

                statService.createAllModes(createdUser).bind()

                log.debug(
                    "User registered with username: {} and email: {}",
                    createdUser.username,
                    createdUser.email,
                )

                UserCreated(createdUser)
            }
            .mapError { listOf(it) }
    }
}
